## 2018-19 NBA Regular Season Box Score Data

This data comes from https://www.nbastuffer.com/2018-2019-nba-player-stats/ for the 2018-19 regular season.

In [1]:
import pandas as pd

df = pd.read_csv('2018-2019_NBA_Player_Stats.csv')
df

FULL NAME TEAM  POS    AGE  GP   MPG  MIN%  USG%  Tor%  FTA  ...  RPG  \
0    Alex Abrines  Okc    G  25.69  31  19.0  39.5  12.2   7.9   13  ...  1.5   
1      Quincy Acy  Pho    F  28.51  10  12.3  25.6   9.2  15.2   10  ...  2.5   
2    Jaylen Adams  Atl    G  22.93  34  12.6  26.2  13.5  19.7    9  ...  1.8   
3    Steven Adams  Okc    C  25.73  80  33.4  69.5  16.4  12.8  292  ...  9.5   
4     Bam Adebayo  Mia  C-F  21.73  82  23.3  48.6  15.8  17.1  226  ...  7.3   
..            ...  ...  ...    ...  ..   ...   ...   ...   ...  ...  ...  ...   
617  Tyler Zeller  Atl    C  29.23   2   5.4  11.1   7.6   0.0    0  ...  3.0   
618  Tyler Zeller  Mem    C  29.23   4  20.6  42.8  21.7  10.0   18  ...  4.5   
619    Ante Zizic  Cle    C  22.27  59  18.3  38.2  18.2  13.8  132  ...  5.4   
620   Ivica Zubac  Lal    C  22.07  33  15.6  32.6  20.5  12.9   66  ...  4.9   
621   Ivica Zubac  Lac    C  22.07  26  20.1  41.9  20.1  14.8   60  ...  7.7   

     TRB%  APG  AST%   SPG   BPG  TOPG   VI   ORTG   DRTG  
0     4.2  0.6   4.3  0.55  0.19  0.45  3.5  103.1  103.9  
1    11.3  0.8   8.2  0.10  0.40  0.40  4.9   87.1   98.5  
2     7.5  1.9  20.2  0.38  0.15  0.82  7.0   99.5  108.1  
3    14.7  1.6   6.6  1.49  0.96  1.73  7.1  119.9  102.7  
4    16.6  2.2  14.2  0.88  0.79  1.48  9.0  120.0   97.2  
..    ...  ...   ...   ...   ...   ...  ...    ...    ...  
617  29.5  0.5  10.9  0.00  0.00  0.00  0.0   82.4   74.1  
618  12.3  0.8   6.2  0.25  0.75  1.00  6.7  126.8   93.4  
619  16.4  0.9   7.7  0.22  0.37  1.05  7.3  116.2  108.1  
620  16.3  0.8   7.3  0.15  0.82  1.00  8.3  120.7   97.9  
621  20.3  1.5  11.0  0.38  0.92  1.42  9.5  114.0   99.1  

[622 rows x 28 columns]

The first thing to notice is that the CSV file includes 622 rows, and some players appear more than once. We can confirm this by running `value_counts()` on the `FULL NAME` column.

In [2]:
df['FULL NAME'].value_counts()

Wesley Matthews    3
Andrew Harrison    3
Jason Smith        3
Isaiah Canaan      3
Greg Monroe        3
                  ..
Jordan Loyd        1
DeAndre' Bembry    1
Davon Reed         1
Julian Washburn    1
Jordan McRae       1
Name: FULL NAME, Length: 530, dtype: int64

In [3]:
len(df['FULL NAME'].unique())

530

In [4]:
df[df['FULL NAME'] == 'Jason Smith']

FULL NAME TEAM  POS    AGE  GP   MPG  MIN%  USG%  Tor%  FTA  ...  RPG  \
535  Jason Smith  Was  F-C  33.11  12  10.8  22.5  16.1  14.2   12  ...  3.1   
536  Jason Smith  Mil  F-C  33.11   6   6.8  14.1  19.5  26.5    2  ...  1.8   
537  Jason Smith  Nor  F-C  33.11   2  10.2  21.1  22.0   9.2    2  ...  2.0   

     TRB%  APG  AST%   SPG   BPG  TOPG   VI   ORTG   DRTG  
535  15.5  1.0  12.3  0.08  0.42  0.58  8.4  110.4  100.0  
536  13.8  0.2   3.1  0.33  0.33  0.83  5.4   80.0   91.0  
537  10.2  0.5   6.1  0.00  0.00  0.50  6.2   93.3  104.0  

[3 rows x 28 columns]

We notice that players who switched teams during the season have their stats recorded separately for each team. We may wish to combine these separate entries by using a weighted average using the `GP` column.

In [5]:
df.isnull().sum()

FULL NAME     0
TEAM          0
POS           0
AGE           0
GP            0
MPG           0
MIN%          0
USG%          0
Tor%          6
FTA           0
FT%           0
2PA           0
2P%           0
3PA           0
3P%           0
eFG%          6
TS%           6
PPG           0
RPG           0
TRB%          0
APG           0
AST%          0
SPG           0
BPG           0
TOPG          0
VI            0
ORTG         25
DRTG         23
dtype: int64

In [6]:
df[df['eFG%'].isnull()]

FULL NAME TEAM  POS    AGE  GP  MPG  MIN%  USG%  Tor%  FTA  ...  RPG  \
267   John Holland  Cle    G  30.43   1  0.7   1.5   0.0   NaN    0  ...  0.0   
326    George King  Pho    F  25.24   1  5.9  12.3   0.0   NaN    0  ...  1.0   
419  Eric Moreland  Pho  F-C  27.30   1  5.0  10.4   0.0   NaN    0  ...  3.0   
529   Kobi Simmons  Cle    G  21.77   1  1.8   3.8   0.0   NaN    0  ...  0.0   
541   Ray Spalding  Dal    F  22.09   1  0.9   1.9   0.0   NaN    0  ...  0.0   
574     Tyler Ulis  Chi    G  23.26   1  0.8   1.7   0.0   NaN    0  ...  0.0   

     TRB%  APG  AST%  SPG  BPG  TOPG   VI  ORTG   DRTG  
267   0.0  0.0   0.0  0.0  0.0   0.0  0.0   NaN    NaN  
326   9.4  0.0   0.0  0.0  0.0   0.0  0.0   NaN  120.8  
419  33.2  0.0   0.0  0.0  0.0   0.0  0.0   NaN   85.1  
529   0.0  0.0   0.0  0.0  0.0   0.0  0.0   NaN    NaN  
541   0.0  0.0   0.0  0.0  0.0   0.0  0.0   NaN    NaN  
574   0.0  0.0   0.0  0.0  0.0   0.0  0.0   NaN    NaN  

[6 rows x 28 columns]

In [7]:
temp = df[df['ORTG'].isnull()]
(temp['MPG'] * temp['GP']).describe()

count    25.000000
mean     13.916000
std      21.717249
min       0.600000
25%       1.800000
50%       5.200000
75%      14.800000
max      92.500000
dtype: float64

There are a number of rows with missing statistics, such as `eFG%`, `TS%`, and `Tor%`. However, these are a result of very small sample sizes - after all, a player can't have an effective field goal percentage if they did not take any shots. In particular, 25 players have no `ORTG`, but this is mostly for players who played less than 15 minutes, and at most 93 minutes. We may wish to remove all rows where the total number of minutes played is below a certain threshold.

In [8]:
df.describe()

AGE          GP         MPG        MIN%        USG%        Tor%  \
count  622.000000  622.000000  622.000000  622.000000  622.000000  616.000000   
mean    26.782379   41.953376   19.216399   40.032154   18.648392   11.935552   
std      4.155828   26.690558    9.210387   19.189038    6.144748    5.587878   
min     19.420000    1.000000    0.600000    1.300000    0.000000    0.000000   
25%     23.445000   17.000000   12.000000   24.950000   15.000000    8.900000   
50%     26.240000   43.000000   19.050000   39.650000   17.800000   11.500000   
75%     29.570000   68.000000   27.200000   56.675000   22.100000   14.300000   
max     42.210000   82.000000   36.900000   76.900000   67.900000   50.000000   

              FTA         FT%          2PA         2P%  ...         RPG  \
count  622.000000  622.000000   622.000000  622.000000  ...  622.000000   
mean    91.244373    0.688080   226.236334    0.481559  ...    3.559646   
std    116.315807    0.232103   252.176351    0.142881  ...    2.526018   
min      0.000000    0.000000     0.000000    0.000000  ...    0.000000   
25%     14.000000    0.643250    37.000000    0.447250  ...    1.800000   
50%     48.500000    0.750000   136.000000    0.500000  ...    3.000000   
75%    123.000000    0.819000   319.750000    0.549000  ...    4.700000   
max    858.000000    1.000000  1277.000000    1.000000  ...   15.600000   

             TRB%         APG        AST%         SPG         BPG        TOPG  \
count  622.000000  622.000000  622.000000  622.000000  622.000000  622.000000   
mean    10.071704    1.864952   13.228296    0.606801    0.377235    1.038939   
std      5.901874    1.708563    9.452492    0.427205    0.398804    0.776064   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      6.225000    0.725000    6.950000    0.310000    0.110000    0.500000   
50%      8.700000    1.300000   10.650000    0.530000    0.270000    0.860000   
75%     12.700000    2.475000   17.675000    0.847500    0.500000    1.427500   
max     57.400000   10.700000   84.200000    2.400000    2.690000    4.960000   

               VI        ORTG        DRTG  
count  622.000000  597.000000  599.000000  
mean     6.859807  107.977387  104.071452  
std      2.607043   16.208868    6.084942  
min      0.000000   20.200000   66.200000  
25%      5.600000  101.800000  100.650000  
50%      6.800000  108.800000  104.600000  
75%      8.175000  115.400000  107.800000  
max     21.800000  300.000000  124.500000  

[8 rows x 25 columns]

In [9]:
df['GP'].sum()/(30*41)

21.215447154471544

This suggests that 21 players saw the floor in the average regular season game - a reasonable figure.

In [10]:
df['TEAM'].value_counts()

Mem    28
Cle    27
Phi    26
Was    25
Mil    24
Pho    24
Nyk    23
Hou    23
Lac    22
Atl    22
Chi    22
Tor    22
Lal    22
Dal    21
Min    21
Sac    20
Nor    20
Det    20
Bro    19
Okc    18
Den    18
Uta    18
Por    18
Mia    18
Bos    17
Gol    17
Ind    17
Orl    17
Cha    17
San    16
Name: TEAM, dtype: int64

There are significant fluctuations in the number of players per team. Some teams, such as the Memphis Grizzlies, had almost 30 players see minutes played over the course of the season, compared to only 16 players for the San Antonio Spurs. The teams with more players may have played more G-League players on 10-day contracts, or been involved in a larger number of trades.

In [11]:
df.corr()

AGE        GP       MPG      MIN%      USG%      Tor%       FTA  \
AGE   1.000000  0.054348  0.143957  0.143795 -0.054783  0.059740  0.025698   
GP    0.054348  1.000000  0.609283  0.609458  0.097783  0.041553  0.649041   
MPG   0.143957  0.609283  1.000000  0.999994  0.246886 -0.001361  0.670073   
MIN%  0.143795  0.609458  0.999994  1.000000  0.246843 -0.001420  0.670054   
USG% -0.054783  0.097783  0.246886  0.246843  1.000000 -0.069857  0.463372   
Tor%  0.059740  0.041553 -0.001361 -0.001420 -0.069857  1.000000  0.036908   
FTA   0.025698  0.649041  0.670073  0.670054  0.463372  0.036908  1.000000   
FT%   0.176470  0.420594  0.466071  0.466184  0.073225 -0.002657  0.264845   
2PA  -0.006042  0.733883  0.721767  0.721841  0.448267  0.021666  0.905885   
2P%   0.003938  0.339273  0.288297  0.288429  0.108700 -0.043161  0.223190   
3PA   0.115245  0.658888  0.661763  0.661801  0.299037 -0.084505  0.613961   
3P%   0.075594  0.185498  0.270983  0.271144 -0.001266 -0.140333  0.090032   
eFG%  0.075956  0.273817  0.233560  0.233724 -0.109396 -0.119914  0.180231   
TS%   0.096165  0.294156  0.265618  0.265814 -0.086974 -0.133658  0.238309   
PPG   0.089061  0.534158  0.876950  0.876999  0.533351 -0.066067  0.831068   
RPG   0.070058  0.434119  0.657452  0.657349  0.234208  0.074437  0.619025   
TRB% -0.004487 -0.003419 -0.054736 -0.054935  0.161414  0.065006  0.121233   
APG   0.145955  0.419740  0.685156  0.685158  0.359849  0.195686  0.612238   
AST%  0.095451  0.144417  0.257548  0.257480  0.383714  0.169853  0.335410   
SPG   0.082976  0.413499  0.743462  0.743537  0.219529  0.113747  0.512508   
BPG  -0.000305  0.358831  0.443343  0.443343  0.098227  0.061908  0.442733   
TOPG  0.071877  0.479619  0.774644  0.774638  0.498527  0.336861  0.775782   
VI    0.089516  0.330056  0.410670  0.410674  0.434271  0.137796  0.514656   
ORTG  0.109182  0.269476  0.212205  0.212381 -0.025324 -0.315579  0.228026   
DRTG -0.026136  0.062033  0.264221  0.264273  0.019745 -0.202145  0.043969   

           FT%       2PA       2P%  ...       RPG      TRB%       APG  \
AGE   0.176470 -0.006042  0.003938  ...  0.070058 -0.004487  0.145955   
GP    0.420594  0.733883  0.339273  ...  0.434119 -0.003419  0.419740   
MPG   0.466071  0.721767  0.288297  ...  0.657452 -0.054736  0.685156   
MIN%  0.466184  0.721841  0.288429  ...  0.657349 -0.054935  0.685158   
USG%  0.073225  0.448267  0.108700  ...  0.234208  0.161414  0.359849   
Tor% -0.002657  0.021666 -0.043161  ...  0.074437  0.065006  0.195686   
FTA   0.264845  0.905885  0.223190  ...  0.619025  0.121233  0.612238   
FT%   1.000000  0.283386  0.321417  ...  0.247794 -0.083493  0.300931   
2PA   0.283386  1.000000  0.238687  ...  0.637459  0.108325  0.624143   
2P%   0.321417  0.238687  1.000000  ...  0.347568  0.141272  0.121819   
3PA   0.362481  0.596420  0.117585  ...  0.227476 -0.222285  0.531612   
3P%   0.278110  0.105968 -0.007417  ... -0.025294 -0.303837  0.205037   
eFG%  0.164338  0.191915  0.569197  ...  0.262667  0.005591  0.072622   
TS%   0.289761  0.228885  0.534557  ...  0.291143  0.019082  0.115527   
PPG   0.408542  0.833254  0.289753  ...  0.660341  0.032107  0.687194   
RPG   0.247794  0.637459  0.347568  ...  1.000000  0.554967  0.354197   
TRB% -0.083493  0.108325  0.141272  ...  0.554967  1.000000 -0.132060   
APG   0.300931  0.624143  0.121819  ...  0.354197 -0.132060  1.000000   
AST%  0.121981  0.330308  0.064644  ...  0.072003 -0.075279  0.744706   
SPG   0.281564  0.536216  0.211679  ...  0.459247 -0.078099  0.658402   
BPG   0.165168  0.452781  0.333653  ...  0.705043  0.379270  0.148989   
TOPG  0.318997  0.764176  0.200249  ...  0.608212  0.050016  0.827403   
VI    0.336906  0.511690  0.364196  ...  0.547375  0.352183  0.579593   
ORTG  0.247073  0.212855  0.412550  ...  0.308339  0.221036  0.113756   
DRTG  0.122542  0.082045 -0.187623  ... -0.204830 -0.596321  0.203292   

          AST%       SPG       BPG      TOPG        VI      ORTG      DRTG  
AGE   

This matrix, although overwhelming, shows us several quick observations. For instance, `MPG` is (obviously) highly correlated with volume metrics such as `PPG`. Interestingly, `USG%` and `eFG%` have a negative correlation.